### PR TITLE
refactor: use coin arrays

### PR DESCRIPTION
**Pull request by cosmic-vagabond**
## Description

Closes:
* https://github.com/elys-network/issues/issues/449

### What has Changed?

* Use Coin arrays instead of separate string and integer to define a Coin
* removed unecssary fields from MsgClose
* Fix test cases
* Fix hook name typo

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/elys-network/elys/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#testing)
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### Deployment Notes

Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.

### Screenshots and Videos

Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.

### CODE REVIEW

File: x/margin/keeper/close.go [hunk 1]
```diff
@@ -31,20 +31,22 @@ func (k Keeper) Close(ctx sdk.Context, msg *types.MsgClose) (*types.MsgCloseResp
 		return nil, sdkerrors.Wrap(types.ErrInvalidPosition, mtp.Position.String())
 	}
 
+	collateralIndex := len(mtp.Collaterals) - 1
```
Comment (jelysn) [inline x/margin/keeper/close.go:34]: Is there any reason using last one for events?

File: x/margin/keeper/open_long_test.go [hunk 2]
```diff
@@ -64,6 +64,8 @@ func TestOpenLong_PoolDisabled(t *testing.T) {
 		msg = &types.MsgOpen{
 			Leverage:         math.LegacyNewDec(10),
 			CollateralAmount: math.NewInt(1),
+			CollateralAsset:  "aaa",
+			BorrowAsset:      "bbb",
```
Comment (jelysn) [inline x/margin/keeper/open_long_test.go:68]: Could you use meaningful assets lik "ATOM" or "USDC" here? "aaa" and "bbb" seems not good even it's a test case.

File: x/margin/types/add_or_append_coin.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package types
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// Utility function to add a new coin or aggregate the amount if coin with same denom already exists
+func AddOrAppendCoin(coins []sdk.Coin, coin sdk.Coin) []sdk.Coin {
```
Comment (jelysn) [inline x/margin/types/add_or_append_coin.go:6]: I think this function is not required, `sdk.Coins(coins).Add(coin)` works exactly same as what this function does

File: x/margin/keeper/close.go (hunk 1 above)
Comment (cosmic-vagabond) [reply x/margin/keeper/close.go:34]: @kenta-elys can you comment on that, I reproduced what you did for open events

File: x/margin/keeper/open_long_test.go (hunk 2 above)
Comment (cosmic-vagabond) [reply x/margin/keeper/open_long_test.go:68]: I do for the following test cases

File: x/margin/types/add_or_append_coin.go (hunk 3 above)
Comment (cosmic-vagabond) [reply x/margin/types/add_or_append_coin.go:6]: @jelysn not really the purpose of this function is to add an asset only if it does exist yet in `[]sdk.Coin`

Comment (jelysn) [reply x/margin/types/add_or_append_coin.go:6]: I don't see other functionality it provide other than sdk.Coins.Add function - can you check sdk.Coins.Add function and check difference?

Comment (cosmic-vagabond) [reply x/margin/types/add_or_append_coin.go]: @jelysn updated the code to use `Coin.Add` good call out

File: x/margin/keeper/close.go (hunk 1 above)
Comment (cosmic-vagabond) [reply x/margin/keeper/close.go:34]: @jelysn @kenta-elys I will add a change in my next PR to include all the denoms in the event attributes

**Pull request merged**